samples/demand_paging/src/main.c

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (c) 2024 BayLibre SAS
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/linker/sections.h>
+
+#include <mmu.h> /* for k_mem_num_pagefaults_get() */
+
+/*
+ * We want to artificially create a huge body of code hence the volatile
+ * to prevent the compiler from optimizing everything down.
+ */
+volatile long foo_val1;
+volatile long foo_val2;
+
+/*
+ * This function is tagged with __ondemand_func to be placed in a special
+ * "ondemand" segment by the linker. This means it is not loaded into memory
+ * at boot time but rather page-by-page on demand when actually executed.
+ * It is also subject to being evicted when memory reclamation occurs.
+ */
+static void __ondemand_func huge_evictable_function(void)
+{
+	foo_val1 = 13131313;
+	foo_val2 = 45454545;
+
+#define CODE0 foo_val1 *= foo_val2; foo_val2 += 1234567; foo_val1 -= 9876543210;
+#define CODE1 CODE0 CODE0 CODE0 CODE0 CODE0 CODE0 CODE0 CODE0 CODE0 CODE0
+#define CODE2 CODE1 CODE1 CODE1 CODE1 CODE1 CODE1 CODE1 CODE1 CODE1 CODE1
+#define CODE3 CODE2 CODE2 CODE2 CODE2 CODE2 CODE2 CODE2 CODE2 CODE2 CODE2
+#define CODE4 CODE3 CODE3 CODE3 CODE3 CODE3 CODE3 CODE3 CODE3 CODE3 CODE3
+#define CODE5 CODE4 CODE4 CODE4 CODE4 CODE4 CODE4 CODE4 CODE4 CODE4 CODE4
+
+	/* This is 100000 repetitions of CODE0 */
+	CODE5
+
+#undef CODE5
+#undef CODE4
+#undef CODE3
+#undef CODE2
+#undef CODE1
+#undef CODE0
```

Comment:
not necessarily a request for change but perhaps it would be more readable / self-explanatory like this? sadly, this is not any faster :)

```suggestion
#define CODE_REPEAT_10(x)     x x x x x x x x x x
#define CODE_REPEAT_100(x)    CODE_REPEAT_10(CODE_REPEAT_10(x))
#define CODE_REPEAT_1000(x)   CODE_REPEAT_10(CODE_REPEAT_100(x))
#define CODE_REPEAT_10000(x)  CODE_REPEAT_10(CODE_REPEAT_1000(x))
#define CODE_REPEAT_100000(x) CODE_REPEAT_10(CODE_REPEAT_10000(x))

	CODE_REPEAT_100000(CODE0)

#undef CODE0
```